Main.hs

```diff
@@ -106,38 +106,29 @@ convertMove m s r
    | m `elem` "rR" = Discard (convertCard s r)
    | otherwise     = error "Unrecognized move"
 
-readCards :: IO [Card]
-readCards = readCards' []
-   where
-   readCards' :: [Card] -> IO [Card]
-   readCards' cs = do
-      s <- getChar
-      if s == '.'
-         then do getChar -- to get new line character
-                 return cs
-         else do r <- getChar
-                 getChar -- to get new line character
-                 let c = convertCard s r
-                 readCards' (cs ++ [c])
+readCards :: IO CardList
+readCards = do line <- getLine
+               if line == "."
+                  then return []
+                  else do rest <- readCards
+                          return ([validateAndConvertCard line] ++ rest) where
+                            validateAndConvertCard :: String -> Card
```

Comment:
When function name becomes long, you can see it as a sign of overcomplicated function. `validateAndConvertCard` is a long name for a function I think. It is a generally good idea to give only one job to a function (it has two jobs now). So maybe we should split this function into two functions: `validate` and `convertCard`. Luckily, we already have `convertCard` function :) and the job of `validate` function should be returning right characters for good inputs or raising an error for bad inputs. Then, we can use two functions as `(convertCard . validate)` which is a better solution in my opinion. What do you think?